src/hotspot/share/runtime/continuationFreezeThaw.cpp

```diff
@@ -507,108 +535,93 @@ bool Freeze<ConfigT>::is_chunk_available_for_fast_freeze(intptr_t* frame_sp
-
-    assert(is_chunk_available0, "");
-
-    if (chunk->sp() < chunk->stack_size()) { // we are copying into a non-empty chunk
-      DEBUG_ONLY(empty = false;)
-      assert(chunk->sp() < (chunk->stack_size() - chunk->argsize()), "");
-      assert(*(address*)(chunk->sp_address() - frame::sender_sp_ret_address_offset()) == chunk->pc(), "");
-
-      chunk_start_sp = chunk->sp() + _cont.argsize(); // we overlap; we'll overwrite the chunk's top frame's callee arguments
-      assert(chunk_start_sp <= chunk->stack_size(), "sp not pointing into stack");
-
-      // increase max_size by what we're freezing minus the overlap
-      chunk->set_max_size(chunk->max_size() + cont_size - _cont.argsize());
-
-      intptr_t* const bottom_sp = cont_stack_bottom - _cont.argsize();
-      assert(bottom_sp == _bottom_address, "");
-      // Because the chunk isn't empty, we know there's a caller in the chunk, therefore the bottom-most frame
-      // should have a return barrier (installed back when we thawed it).
-      assert(*(address*)(bottom_sp-frame::sender_sp_ret_address_offset()) == StubRoutines::cont_returnBarrier(),
-             "should be the continuation return barrier");
-      // We copy the fp from the chunk back to the stack because it contains some caller data,
-      // including, possibly, an oop that might have gone stale since we thawed.
-      patch_stack_pd(bottom_sp, chunk->sp_address());
-      // we don't patch the return pc at this time, so as not to make the stack unwalkable for async walks
-    } else { // the chunk is empty
-      chunk_start_sp = chunk->sp();
-
-      assert(chunk_start_sp == chunk->stack_size(), "");
-
-      chunk->set_max_size(cont_size);
-      chunk->set_argsize(_cont.argsize());
-    }
-  } else { // no chunk; allocate
-    assert(_thread->thread_state() == _thread_in_vm, "");
-    assert(!is_chunk_available_for_fast_freeze(frame_sp), "");
-    assert(_thread->cont_fastpath(), "");
-
-    chunk = allocate_chunk(cont_size + frame::metadata_words);
-    if (UNLIKELY(chunk == nullptr || !_thread->cont_fastpath() || _barriers)) { // OOME/probably humongous
-      log_develop_trace(continuations)("Retrying slow. Barriers: %d", _barriers);
-      return false;
-    }
+  assert(_thread->thread_state() == _thread_in_vm, "");
+  assert(_is_chunk_available_size > 0, "");
+  assert(_thread->cont_fastpath(), "");
 
-    chunk->set_max_size(cont_size);
-    chunk->set_argsize(_cont.argsize());
+  if (UNLIKELY(chunk == nullptr || !_thread->cont_fastpath() || _barriers)) { // OOME/probably humongous
+    log_develop_trace(continuations)("Retrying slow. Barriers: %d", _barriers);
+    return false;
+  }
 
-    // in a fresh chunk, we freeze *with* the bottom-most frame's stack arguments.
-    // They'll then be stored twice: in the chunk and in the parent chunk's top frame
-    chunk_start_sp = cont_size + frame::metadata_words;
-    assert(chunk_start_sp == chunk->stack_size(), "");
+  chunk->set_max_size(_cont_size);
```

Comment:
This variable _cont_size is really strange.  Above we add in argsize for total size that we need to freeze, but here we set max_size to cont_size.  We should go through and see what we add and subtract from cont_size in various places and see what size is what (and if it's easier to figure out what's correct).  After this patch.
I assume that this patch is mostly code motion and doesn't change the values in places, which is good.

Reply:
Yes